Script to add rapiddisk to initrd images and use it on boot device

Shub77 opened the pull request:
Hi there,

I wrote a script which installs all the needed scripts to make update-initramfs or dracut to include Rapiddisk's modules and executable into the initrd file:

### Ubuntu

```
shub@gturismo:/mnt/compile/rapiddisk/misc$ sudo ./install_initrd.sh --boot=/dev/sda1 --size=100 --kernel=`uname -r` --force
 - Copying /mnt/compile/rapiddisk/misc/hooks/rapiddisk_hook to /etc/initramfs-tools/hooks/rapiddisk_hook...
 - Copying /mnt/compile/rapiddisk/misc/scripts/init-premount/rapiddisk to /etc/initramfs-tools/scripts/init-premount/rapiddisk...
 - Updating initramfs for kernel version 5.4.0-31-generic...

update-initramfs: Generating /boot/initrd.img-5.4.0-31-generic

 - Updating grub...

Sourcing file `/etc/default/grub'
Sourcing file `/etc/default/grub.d/init-select.cfg'
Generating grub configuration file ...
Found linux image: /boot/vmlinuz-5.4.0-31-generic
Found initrd image: /boot/initrd.img-5.4.0-31-generic
Found linux image: /boot/vmlinuz-5.4.0-29-generic
Found initrd image: /boot/initrd.img-5.4.0-29-generic
Found memtest86+ image: /boot/memtest86+.elf
Found memtest86+ image: /boot/memtest86+.bin
Found Ubuntu 20.04 LTS (20.04) on /dev/sda1
done

Done.
shub@gturismo:/mnt/compile/rapiddisk/misc$
```

Atfter a reboot:

    shub@gturismo:/mnt/compile/rapiddisk/misc$ mount | grep -iP 'on / '
    /dev/mapper/rc-wa_sda1 on / type ext4 (rw,relatime,errors=remount-ro)
    shub@gturismo:/mnt/compile/rapiddisk/misc$

### CentOS 8
```
[shub@centos8 misc]$ sudo ./install_initrd.sh --kernel=4.18.0-147.8.1.el8_1.x86_64 --size=100 --boot=/dev/sda2
[sudo] password for shub:
 - CentOS detected!
 - Installing by copying 96rapiddisk into /usr/lib/dracut/modules.d/...
 - Editing /usr/lib/dracut/modules.d/96rapiddisk/run_rapiddisk.sh...
 - chmod +x /usr/lib/dracut/modules.d/96rapiddisk/* ...
 - Running 'dracut --kver 4.18.0-147.8.1.el8_1.x86_64 -f'
 - Done under CentOS.
[shub@centos8 misc]$
```
After a reboot:

    [shub@centos8 ~]$ mount | grep -iP 'on / '
    /dev/mapper/rc-wa_sda2 on / type xfs (rw,relatime,seclabel,attr2,inode64,noquota)
    [shub@centos8 ~]$

Let me know if it works!

Cheers

pkoutoupis commented:
Awesome! I will review this shortly. A couple of notes:
1. Thank you very much for doing this.
2. Instead of placing it into a misc directory, I was thinking that maybe we can instead place this into a misc/boot-device subdirectory? Just in case we decide to add more misc scripts and binaries sometime in the future.
3. Will this require its own independent Makefile? I wouldn't not want to plug it into root's Makefile. At least not yet. Or what about a README.md in misc/boot-device with instructions?

pkoutoupis commented on misc/96rapiddisk/module-setup.sh:
```diff
@@ -0,0 +1,24 @@
+}
+
+installkernel() {
+
```
(style) no big deal and you do not have to but to maintain consistency with other functions, I would remove the blank lines above and below the "hostonly=" line.

pkoutoupis commented on misc/install_initrd.sh:
```diff
@@ -0,0 +1,221 @@
+		rm -f "$hook_dest" 2>/dev/null
+		rm -f "$bootscript_dest" 2>/dev/null
+	fi
+	#if lsinitramfs /boot/initrd.img-"$kernel_version" | grep rapiddisk >/dev/null 2>/dev/null ; then
```
Why is this commented out? Do we still need it here?

Shub77 replied:
Is commented out because I don't know if it really useful... yet!

pkoutoupis commented on misc/install_initrd.sh:
```diff
@@ -0,0 +1,221 @@
+##############################################################################
+# TODO The following piece of code was there to check if the user specified
+# the same device present in /etc/fstab - but if in fstab UUIDs are used, or
+# the LABEL= syntax, it looses meaning.
```
minor minor minor detail: misspelling: supposed to be "loses"

Shub77 replied:
I fixed it!

Shub77 commented:
> Awesome! I will review this shortly. A couple of notes:
> 
>     1. Thank you very much for doing this.

:-) You're welcome, I am happy to help.

>     2. Instead of placing it into a misc directory, I was thinking that maybe we can instead place this into a misc/boot-device subdirectory? Just in case we decide to add more misc scripts and binaries sometime in the future.

Well, one point must be clear: this is the very very first attempt, so the dir structure, the (many) typos/errors or the bad styled code in the bash script are for sure to be changed, improved, corrected! I createted an ubuntu and a centos subdir in the script dir. The dir containing the script, can me named and moved in any place, because now  the refrences to the files are relative to the script's directory.

>     3. Will this require its own independent Makefile? I wouldn't not want to plug it into root's Makefile. At least not yet. Or what about a README.md in misc/boot-device with instructions (such as where/how to run the install script)?

I thought about this, and I choose to write a script because a Makefile has another role. A real Makefile, is supposed to help to build code, this one is a matter of installing files into the system and run a couple of tools.

I saw some Makefiles which did magics (to my eyes at least), and we can call a script from a Makefile. This can be done. But I believe a bash script is necessary anyway, and that using make would have complicated things much.

The README.md will arrive, writing docs is always a bit boring :-) and I tried to write an help with as many infos as I could.

Note: I wrote the script under Ubuntu. Since I understood you're under CentOS, I installed it under CentOS 8 and... find out that the whole initrd creation process has nothing in common with the Ubuntu one!

I added CentOS support and I tested it under CentOS 8, but it should work under 7 too, as long as `dracut` was already there...

I have a new pull request, I don't know how to attach it to this one though!

Cheers

Shub77 commented on misc/install_initrd.sh:
```diff
@@ -0,0 +1,245 @@
+        exit 0
+    fi
+
+    ##############################################################################
+    # TODO The following piece of code was there to check if the user specified
+    # the same device present in /etc/fstab - but if in fstab UUIDs are used, or
+    # the LABEL= syntax, it loses meaning.
+    ##############################################################################
```
If in fstab the device is specified via UUID or LABEL it is possible to check `/dev/disk/by-*` to "resolve" it into the classic `/dev/xxxx` syntx... I don't know if it really makes sense.

pkoutoupis replied:
I think most modern distros tend to use the UUID approach. But if we initially label this as experimental and have a README.md highlighting features yet-to-be integrated, we should be good for an initial merge.

pkoutoupis commented on misc/install_initrd.sh:
```diff
@@ -0,0 +1,245 @@
+    echo ""
+    echo "Where:"
+    echo ""
+    echo "<boot_partition> is the device mounted as '/' as in /etc/fstab in the /dev/xxxn format"
```
There may be a bit of confusion for the user as sometimes the "boot partition" may reside as /boot in its own partition. Just a suggestion to relabel "boot_partition" here and above, alongside the --boot arg as "root_partition" and --root. Thoughts?

Shub77 replied:
Using the "root" word is way better!

pkoutoupis commented:
FYI, I mentioned version 6.1. I want to get this into that version. I think it would be a nice addition.

Shub77 commented:
> I see that some files are still using tabs while others you recently changed to 4 space indentations. Would it be too much trouble to revert back to tabs? I am currently working in a branch for version 6.1 and modified the remaining files using spaces to keep with the tabbing consistency.

I switched to tab indents all of the scripts, np.
